lib/rules/no-unused-vars.js

```diff
@@ -832,6 +836,607 @@ module.exports = {
+                    return fixVariables(parentNode);
+                }
+
+                // remove last unused array element
+                if (
+                    getTokenBeforeValue(node) === "," &&
+                        getTokenAfterValue(node) === "]"
+                ) {
+                    return fixer.removeRange([getPreviousTokenStart(node), node.range[1]]);
+                }
+
+                // remove unused array element
+                return fixer.removeRange(node.range);
+            }
+
+            /**
+             * fix cases like {a: {k}} or {a: [k]}
+             * @param {ASTNode} node parent node to check
+             * @returns {Object} fixer object
+             */
+            function fixObjectWithValueSeperator(node) {
+                const parentNode = node.parent.parent;
+
+                // fix cases like [{a : { b }}] or [{a : [ b ]}]
+                if (
+                    parentNode.parent.type === "ArrayPattern" &&
+                        parentNode.properties.length === 1
+                ) {
+                    return fixNestedArrayVariable(parentNode);
+                }
+
+                // fix cases like {a: {k}} or {a: [k]}
+                return fixNestedObjectVariable(node);
+            }
+
+            /**
+             * fix ...[[a]] or ...[{a}] like patterns
+             * @param {ASTNode} node parent node to check
+             * @returns {Object} fixer object
+             */
+            function fixRestInPattern(node) {
+                const parentNode = node.parent;
+
+                // fix ...[[a]] or ...[{a}] in function parameters
+                if (isFunction(parentNode)) {
+                    if (parentNode.params.length === 1) {
+                        return fixer.removeRange(node.range);
+                    }
+
+                    return fixer.removeRange([getPreviousTokenStart(node), node.range[1]]);
+                }
+
+                // fix rest in nested array pattern like [[a, ...[b]]]
+                if (parentNode.type === "ArrayPattern") {
+
+                    // fix [[...[b]]]
+                    if (hasSingleElement(parentNode)) {
+                        if (parentNode.parent.type === "ArrayPattern") {
+                            return fixNestedArrayVariable(parentNode);
+                        }
+
+                        // fix 'const [...[b]] = foo; and function foo([...[b]]) {}
+                        return fixVariables(parentNode);
+                    }
+
+                    // fix [[a, ...[b]]]
+                    return fixer.removeRange([getPreviousTokenStart(node), node.range[1]]);
+                }
+
+                return null;
+            }
+
+            // skip fix when variable is reassigned
+            if (writeReferences.length > 1 && parentType !== "AssignmentPattern") {
+                return null;
+            }
```

Comment:
Variables initialized in an assignment pattern would be removed even if they have write references, which is probably not what we want:

```js
let [a = 1] = arr; // suggestion removes this

a = 2;
```

Reply:
@mdjermanovic, should we skip suggestion here also for `b`?
```js
/*eslint no-unused-vars: ["error", { "destructuredArrayIgnorePattern": "^_" }]*/

let _a, b; foo.forEach(item => { [a, b] = item; });

//'_a' is defined but never used.
//'b' is assigned a value but never used.
```

Reply:
If the suggestion is going to remove only the declaration, then yes we should skip the suggestion because it would leave unresolved references in the code.

Reply:
Done! thanks.